File: 0429_levelOrder.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

from collections import deque

class Solution:
    def levelOrder(self, root: 'Node') -> List[List[int]]:
        if not root: return []
        Q = deque([[root]])
        output = [[root.val]]
        while Q:
            node_list = Q.popleft()
            temp = []
            nodes = []
            for node in node_list:
                for child in node.children:
                    nodes.append(child)
                    temp.append(child.val)
            if temp:
                output.append(temp)
            if nodes:
                Q.append(nodes)
        return output

# Alternative solution

class Solution:
    def levelOrder(self, root: 'Node') -> List[List[int]]:
        if not root:
            return []
        queue = deque([root])
        levelOrder = []
        while queue:
            levelSize = len(queue)
            level = []
            for _ in range(levelSize):
                currentNode = queue.popleft()
                level.append(currentNode.val)
                for node in currentNode.children:
                    queue.append(node)
            levelOrder.append(level)
        return levelOrder

File: 0001-0500/0429_levelOrder.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<List<Integer>> levelOrder(Node root) {
        List<List<Integer>> output = new ArrayList<>();

        if (root == null) return output;

        Queue<Node> queue = new LinkedList<>() {{add(root);}};

        while (!queue.isEmpty()) {
            int length = queue.size();
            List<Integer> level = new ArrayList<>();

            for (int i = 0; i < length; i++) {
                Node node = queue.remove();
                level.add(node.val);

                for (Node child: node.children) {
                    queue.add(child);
                }
            }
            output.add(level);
        }

        return output;
    }
}
